frontend/packages/dev-console/src/components/pipelineruns/logs/logs-utils.ts

```diff
@@ -26,3 +44,119 @@ export const getRenderContainers = (
     stillFetching: firstRunningCont !== -1,
   };
 };
+
+const getOrderedStepsFromPod = (name: string, ns: string): Promise<ContainerStatus[]> => {
+  return k8sGet(PodModel, name, ns)
+    .then((pod: PodKind) => {
+      return getSortedContainerStatus(
+        pod.spec.containers ?? [],
+        pod.status?.containerStatuses ?? [],
+      );
+    })
+    .catch((err) => {
+      errorModal({ error: err.message || 'Error downloading logs.' });
+      return [];
+    });
+};
+
+type StepsWatchUrl = {
+  [key: string]: {
+    name: string;
+    steps: { [step: string]: WatchURLStatus };
+  };
+};
+
+export type DownloadAll = {
+  fileName: string;
+  logs: string;
+};
+
+type WatchURLStatus = {
+  status: string;
+  url: string;
+};
+
+export const getDownloadAllLogsCallback = (
+  sortedTaskRuns: string[],
+  taskRunFromYaml: TaskRuns,
+  namespace: string,
+  pipelineRunName: string,
+): (() => Promise<DownloadAll>) => {
+  const getWatchUrls = async (): Promise<StepsWatchUrl> => {
+    const stepsList: ContainerStatus[][] = await Promise.all(
+      sortedTaskRuns.map((currTask) => {
+        const { status = {} } = taskRunFromYaml[currTask];
+        const { podName } = status;
+        return getOrderedStepsFromPod(podName, namespace);
+      }),
+    );
+    return sortedTaskRuns.reduce((acc, currTask, i) => {
+      const { pipelineTaskName, status = {} } = taskRunFromYaml[currTask];
+      const { podName } = status;
+      const steps = stepsList[i];
+      const allStepUrls = steps.reduce((stepUrls, currentStep) => {
+        const { name } = currentStep;
+        const currentStatus = containerToLogSourceStatus(currentStep);
+        if (currentStatus === LOG_SOURCE_WAITING) return stepUrls;
+        const urlOpts = {
+          ns: namespace,
+          name: podName,
+          path: 'log',
+          queryParams: {
+            container: name,
+            follow: 'true',
+          },
+        };
+        return {
+          ...stepUrls,
+          [name]: {
+            status: currentStatus,
+            url: resourceURL(PodModel, urlOpts),
+          } as WatchURLStatus,
+        };
+      }, {});
+      acc[currTask] = {
+        name: pipelineTaskName,
+        steps: { ...allStepUrls },
+      };
+      return acc;
+    }, {});
+  };
+
+  const fetchLogs = async (tasksPromise: Promise<StepsWatchUrl>) => {
+    const tasks = await tasksPromise;
+    const allRequests: Promise<string>[] = sortedTaskRuns.reduce((acc, currTask) => {
+      const task = tasks[currTask];
+      const promises: Promise<string>[] = Object.keys(task.steps).map((step, i) => {
+        let heading = '';
+        if (i === 0) {
+          heading += `${task.name}\n\n`;
```

Comment:
why not simply initialize the heading with taskName

Reply:
I'm confused? so on `i === 1` we get the header again? This is in a loop.

Reply:
Oh, outside of the "block" as in the `.map`. That'd work too.

Reply:
Oh no that won't... Everything is internal to this function now... what are you referring to @abhinandan13jan ?

Reply:
It's ok to keep it... I might be missing on something... Cant the taskName be prepended during return. (i===0) check is not the best possible way...

Reply:
Ah, that is true. That could be done. If we add it on 138 instead of here... that's a good point. Not going to be a huge problem; Sahil needs to work to get this merged and this is tiny. I think it's okay to leave it how it is.